packages/go_router/example/lib/stateful_shell_route.dart

```diff
@@ -0,0 +1,523 @@
+                ),
+                routes: <RouteBase>[
+                  GoRoute(
+                    path: 'details/:param',
+                    builder: (BuildContext context, GoRouterState state) =>
+                        DetailsScreen(
+                      label: 'B',
+                      param: state.params['param'],
+                      extra: state.extra,
+                    ),
+                  ),
+                ],
+              ),
+            ],
+          ),
+
+          /// The route branch for the third tab of the bottom navigation bar.
+          StatefulShellBranch(
+            /// StatefulShellBranch will automatically use the first descendant
+            /// GoRoute as the initial location of the branch. If another route
+            /// is desired, specify the location of it using the defaultLocation
+            /// parameter.
+            // defaultLocation: '/c2',
+            routes: <RouteBase>[
+              StatefulShellRoute(
+                /// This bottom tab uses a nested shell, wrapping sub routes in a
+                /// top TabBar.
+                branches: <StatefulShellBranch>[
+                  StatefulShellBranch(routes: <GoRoute>[
+                    GoRoute(
+                      path: '/c1',
+                      builder: (BuildContext context, GoRouterState state) =>
+                          const TabScreen(
+                              label: 'C1', detailsPath: '/c1/details'),
+                      routes: <RouteBase>[
+                        GoRoute(
+                          path: 'details',
+                          builder:
+                              (BuildContext context, GoRouterState state) =>
+                                  DetailsScreen(
+                            label: 'C1',
+                            extra: state.extra,
+                            withScaffold: false,
+                          ),
+                        ),
+                      ],
+                    ),
+                  ]),
+                  StatefulShellBranch(routes: <GoRoute>[
+                    GoRoute(
+                      path: '/c2',
+                      builder: (BuildContext context, GoRouterState state) =>
+                          const TabScreen(
+                              label: 'C2', detailsPath: '/c2/details'),
+                      routes: <RouteBase>[
+                        GoRoute(
+                          path: 'details',
+                          builder:
+                              (BuildContext context, GoRouterState state) =>
+                                  DetailsScreen(
+                            label: 'C2',
+                            extra: state.extra,
+                            withScaffold: false,
+                          ),
+                        ),
+                      ],
+                    ),
+                  ]),
+                ],
+                builder: (StatefulShellBuilder shellBuilder) {
```

Comment:
Well, somewhere the StatefulShellRouteState must still be maintained and managed though (note that the lifetime of StatefulShellRouteState is longer than that of for instance GoRouterState, which is recreated on each call to RouteBuilder). And this means that we need some widget that is placed above ("wrapping") the widget containing the user implementation of the shell body/content. And preferably the user shouldn't be bothered with the details of creating the wrapper widget, which is what the StatefulShellBuilder is trying to solve. 

I mentioned this as response to your comment in the previous review, where I also included below image to try to explain how the pieces fit together:

![StatefulShellRoute-architecture](https://user-images.githubusercontent.com/1228601/219873628-fb553c89-3ea8-4055-b4d4-4d91237bcad2.svg)

Reply:
And regarding the overlapping routes and if that would be ok for branches - good question. Not sure if it perhaps would lead to unexpected behaviour for the user in some cases (i.e. ending up in unexpected branch). I'm wondering if it perhaps would be good to restrict root routes in branches to have a unique path/sub-path prefix (i.e. no overlap of base path over branches)?